File: go/Polygon.go
package main

import (
	"fmt"
	"math"
)

type Polygon struct {
	points []Point
	bBox   *Rectangle
}

//---------------------------------------------------------------------------------------
func (thisPt *Polygon) PointIsInside(point *Point) bool {

	count := 0

	if thisPt.bBox != nil && !thisPt.bBox.PointInside(point) {
		return false
	}

	size := len(thisPt.points)
	for i := 1; i < size; i++ {
		p1 := &thisPt.points[i-1]
		p2 := &thisPt.points[i]

		if p1.GetX() < point.GetX() && p2.GetX() < point.GetX() {
			continue
		}

		if p1.GetY() <= point.GetY() && p2.GetY() >= point.GetY() {
			count++
		} else if p1.GetY() >= point.GetY() && p2.GetY() <= point.GetY() {
			count++
		}
	}
	return ((count % 2) != 0)
}

//---------------------------------------------------------------------------------------
func (thisPt *Polygon) SetPoints(pointList []Point) {
	thisPt.points = pointList
}

//---------------------------------------------------------------------------------------

func (thisPt *Polygon) GetBoundingBox() *Rectangle {
	if thisPt.bBox != nil {
		return thisPt.bBox
	}

	bBox := new(Rectangle)

	start := Point{}
	end := Point{}

	start.SetX(math.MaxFloat64)
	start.SetY(math.MaxFloat64)
	end.SetX(0)
	end.SetY(0)

	for i := range thisPt.points {
		p := &thisPt.points[i]

		if p.GetX() < start.GetX() {
			start.SetX(p.GetX())
		}

		if p.GetY() < start.GetY() {
			start.SetY(p.GetY())
		}

		if p.GetX() > end.GetX() {
			end.SetX(p.GetX())
		}

		if p.GetY() > end.GetY() {
			end.SetY(p.GetY())
		}
	}

	bBox.SetEnd(&end)
	bBox.SetStart(&start)

	thisPt.bBox = bBox
	return thisPt.bBox
}

//---------------------------------------------------------------------------------------
func (thisPt *Polygon) GetWKTString() string {
	out := "POLYGON  (("
	for i, p := range thisPt.points {
		if i > 0 {
			out += ","
		}
		out += fmt.Sprintf("%f %f", p.GetX(), p.GetY())
	}
	out += "))"
	return out
}

//---------------------------------------------------------------------------------------

File: java/src/main/java/geospat/Polygon.java
package geospat;

import java.util.ArrayList;

public class Polygon {

    private ArrayList<Point> points_;
    private Rectangle bBox_=null;
    //-----------------------------------------------------------------------------------
    Polygon(ArrayList<Point> points){
        points_=points;
    }
    //-----------------------------------------------------------------------------------
    Polygon(){}

    //-----------------------------------------------------------------------------------
    Rectangle getBoundingBox(){

    if(bBox_!=null){
      return bBox_;
    }

    Point start = new Point(Float.MAX_VALUE,Float.MAX_VALUE);
    Point end = new Point(-Float.MAX_VALUE,-Float.MAX_VALUE);
    
    for (Point i : points_) {
      if (i.getX() < start.getX()) {
        start.setX(i.getX());
      }

      if (i.getY() < start.getY()) {
        start.setY(i.getY());
      }

      if (i.getX() > end.getX()) {
        end.setX(i.getX());
      }

      if (i.getY() > end.getY()) {
        end.setY(i.getY());
      }
    }
    bBox_ = new Rectangle(start, end);
    return bBox_;
  }
  //-----------------------------------------------------------------------------------
  boolean pointIsInside(final Point point) {
    int count = 0;
    int size=points_.size();

    if(bBox_!=null && !bBox_.pointInside(point)){
      return false;
    }

    for (int i = 1; i < size; i++) {
      final Point p1 = points_.get(i-1);
      final Point p2 = points_.get(i);
      if (p1.getX() < point.getX() && p2.getX() < point.getX()) {
        continue;
      } if (p1.getY() <= point.getY() && p2.getY() >= point.getY()) {
        count++;
      } else if (p1.getY() >= point.getY() && p2.getY() <= point.getY()) {
        count++;
      }
    }
    if ((count % 2)!=0) {
      return true;
    }
    return false;
  }
}
